src/implied_volatility.rs

```diff
@@ -8,117 +12,135 @@ where
+                let f = f * (-self.q * self.t).exp();
+
+                // convert the option type into \theta
+                let q: f64 = match self.option_type {
+                    OptionType::Call => 1.0,
+                    OptionType::Put => -1.0,
+                };
+
+                let sigma = implied_volatility_from_a_transformed_rational_guess(
+                    p as f64,
+                    f as f64,
+                    self.k as f64,
+                    self.t as f64,
+                    q as f64,
+                );
+
+                if sigma.is_nan() || sigma.is_infinite() || sigma < 0.0 {
+                    Err("Implied volatility failed to converge".to_string())?
+                }
+                Ok(sigma)
             }
         }
-        Ok(sigma)
-    }
-
-    /// Calculates the implied volatility of the option.
-    /// # Requires
-    /// s, k, r, t, p
-    /// # Returns
-    /// f64 of the implied volatility of the option.
-    /// # Example:
-    /// ```
-    /// use blackscholes::{Inputs, OptionType, ImpliedVolatility};
-    /// let inputs = Inputs::new(OptionType::Call, 100.0, 100.0, Some(0.2), 0.05, 0.05, 20.0/365.25, None);
-    /// let iv = inputs.calc_rational_iv().unwrap();
-    /// ```
-    ///
-    /// Uses the "Let's be rational" method from ["Let’s be rational" (2016) by Peter Jackel](http://www.jaeckel.org/LetsBeRational.pdf)
-    /// from Jackel's C++ implementation, imported through the C FFI.  The C++ implementation is available at [here](http://www.jaeckel.org/LetsBeRational.7z)
-    /// Per Jackel's whitepaper, this method can solve for the implied volatility to f64 precision in 2 iterations.
-    fn calc_rational_iv(&self) -> Result<f64, String> {
-        // extract price, or return error
-        let p = self.p.ok_or("Option price is required".to_string())?;
-
-        // "let's be rational" works with the forward and undiscounted option price, so remove the discount
-        let rate_inv_discount = (self.r * self.t).exp();
-        let p = p * rate_inv_discount;
-
-        // compute the forward price
-        let f = self.s * rate_inv_discount;
-        // The Black-Scholes-Merton formula takes into account dividend yield by setting S = S * e^{-qt}, do this here with the forward
-        let f = f * (- self.q * self.t).exp();
-        
-        // convert the option type into \theta
-        let q: f64 = match self.option_type {
-            OptionType::Call => 1.0,
-            OptionType::Put => -1.0,
-        };
-
-        let sigma = implied_volatility_from_a_transformed_rational_guess(p as f64, f as f64, self.k as f64, self.t as f64, q as f64);
-
-        if sigma.is_nan() || sigma.is_infinite() || sigma < 0.0 {
-            Err("Implied volatility failed to converge".to_string())?
-        }
-        Ok(sigma)
-    }
+    };
 }
+
+impl_iv!(f32, f32);
```

Comment:
maybe i am wrong, but @claireratigan i think it is possible without macros n generation dedicated impl for each type